### PR TITLE
Invoke-Program - Mask verbose output

### PR DESCRIPTION
**Pull request by potatoqualitee**
Per https://github.com/dataplat/dbatools/discussions/8049

I have not tested this.

### CODE REVIEW

**Comment by andreasjordan** [comment]
I can test this if needed.

**Comment by andreasjordan** [comment]
It works, but only for the SA. The service passwords should also be hidden:

image

**Comment by potatoqualitee** [comment]
I appreciate your tests! Ugh, will this require regex? I guess anything that is `PASSWORD="`

**Comment by andreasjordan** [comment]
Oh, and I think it only works, because in my lab I use the same password for SA and the admin account I use to install it. The `Invoke-Command` does not get the credentials.

So yes, regex...

**Comment by andreasjordan** [comment]
These are the passwords that would need masking (from Install-DbaInstance):
```
            # Now apply credentials
            $execParams += Update-ServiceCredential -Node $configNode -Credential $EngineCredential -AccountName SQLSVCACCOUNT
            $execParams += Update-ServiceCredential -Node $configNode -Credential $AgentCredential -AccountName AGTSVCACCOUNT
            $execParams += Update-ServiceCredential -Node $configNode -Credential $ASCredential -AccountName ASSVCACCOUNT
            $execParams += Update-ServiceCredential -Node $configNode -Credential $ISCredential -AccountName ISSVCACCOUNT
            $execParams += Update-ServiceCredential -Node $configNode -Credential $RSCredential -AccountName RSSVCACCOUNT
            $execParams += Update-ServiceCredential -Node $configNode -Credential $FTCredential -AccountName FTSVCACCOUNT
            $execParams += Update-ServiceCredential -Node $configNode -Credential $PBEngineCredential -AccountName PBENGSVCACCOUNT -PasswordName PBDMSSVCPASSWORD
            $execParams += Update-ServiceCredential -Credential $SaCredential -PasswordName SAPWD
```

They are all build this way:
```
"/$PasswordName=`"" + $Credential.GetNetworkCredential().Password + '"'
```

**Comment by potatoqualitee** [comment]
awesome! so far, it looks like a catchall for PASSWORD= would still apply.

**Comment by andreasjordan** [comment]
No, SA is SAPWD.

**Comment by potatoqualitee** [comment]
ahh man, okay, that should be `SAPWD` as a whole, since PWD could mean present working directory

**Comment by andreasjordan** [comment]
This should work: `-replace '(PASSWORD|SAPWD)=".*', '$1="********"'`

**Comment by andreasjordan** [comment]
Yes:
image

**Comment by andreasjordan** [comment]
So line 168 should be:
```
        $secureArgumentList = $ArgumentList -replace '(PASSWORD|SAPWD)=".*', '$1="********"'
```

**Comment by potatoqualitee** [comment]
:O!! hahahahah AWESOME! I was gonna dread this for 48 hours before i finally gave in and did it

File: internal/functions/Invoke-Program.ps1 [hunk 1]
```diff
@@ -164,7 +164,10 @@ function Invoke-Program {
         }
 
         Write-Message -Level Debug -Message "Acceptable success return codes are [$($SuccessReturnCode -join ',')]"
-        Write-Message -Level Verbose -Message "Starting process [$Path] with arguments [$ArgumentList] on $ComputerName through $Authentication protocol"
+
+        $secureArgumentList = "$ArgumentList".Replace($($Credential.GetNetworkCredential().Password), "****************")
```
Comment (potatoqualitee) [inline internal/functions/Invoke-Program.ps1:168]:
```suggestion
                        $secureArgumentList = $ArgumentList -replace '(PASSWORD|SAPWD)=".*', '$1="********"'
```

Comment (potatoqualitee) [reply internal/functions/Invoke-Program.ps1:168]: @andreasjordan please confirm my github suggestion of your comment suggestion 🙇🏼

**Comment by niphlod** [comment]
we're building that string .... just build the secure one along the way instead of trying to solve it with regex .... someone will whine about it not working when the password contains PWD=, quotes, and such. Much easier and streamlined to never have anything "passwordy" on the "securestring"

**Comment by andreasjordan** [comment]
So adding a new parameter to `Invoke-Program`? Would be a much bigger change and more commands involved - but yes, a better solution.

**Comment by potatoqualitee** [comment]
i like this interim solution until make all sorts of bigger changes?

**Comment by niphlod** [comment]
jm2c, is argumentlist passed to invoke-program as a list or as a string ? if it's a list, masking should be easier

**Comment by potatoqualitee** [comment]
we make it into a string with "$ArgumentList"

**Comment by andreasjordan** [comment]
`$ArgumentList` is a list and with `$secureArgumentList = $ArgumentList -replace '(PASSWORD|SAPWD)=".*', '$1="********"'` every item is replaced when needed.

File: internal/functions/Invoke-Program.ps1 (hunk 1 above)
Comment (andreasjordan) [reply internal/functions/Invoke-Program.ps1:168]: Yes, but I think there a too many blanks.

Comment (potatoqualitee) [reply internal/functions/Invoke-Program.ps1:168]: ill commit and see if tests fail 😊

**Review by andreasjordan — APPROVED** [review]
Looks good!

**Comment by potatoqualitee** [comment]
Thanks everyone 💯

**Pull request merged**